Add getSinglerRef function for temporary retrieval of reference datasets

j-andrews7 opened the pull request:
As mentioned in #1, this provides a way to retrieve reference datasets from Github. 

Not quite sure I'm using the `envir` parameter for `load` in quite the way you wanted, but this avoids muddying the `GlobalEnv` and the `eval` tomfoolery I had previously.

I'm willing to add a note about it to the vignette if you'd like.

LTLA commented on R/helper.R:
```diff
@@ -0,0 +1,43 @@
+#'
+#' \code{getSinglerRef} retrieves \code{SingleR} reference datasets from Github. 
+#'
+#' @param dataset String referring to SingleR reference dataset to retrieve. 
```
There's no need to explicitly list the options here, if they are implicitly present in the argument list of the function (see my other comment). I would recommend moving a full description of the options into `@details` along with some plain-English details about what each one is. This information is probably in the old SingleR repo.

LTLA commented on R/helper.R:
```diff
@@ -0,0 +1,43 @@
+#' @export
+#' @importFrom curl curl_download
+#' 
+getSinglerRef <- function(dataset = "hpca") {
```
Move `opts` as the default `dataset=` argument and use `dataset <- match.arg(dataset)` to do the check.

LTLA commented on R/helper.R:
```diff
@@ -0,0 +1,43 @@
+#' mouse.rnaseq <- getSinglerRef(dataset = "mouse.rnaseq")
+#'
+#' @export
+#' @importFrom curl curl_download
```
Switch the imports to pull the relevant functions out of _BiocFileCache_, and don't forget to add _BiocFileCache_ to the `Imports:` of the `DESCRIPTION`.

LTLA commented on R/helper.R:
```diff
@@ -0,0 +1,43 @@
+	  bfc <- BiocFileCache(ask=FALSE)
+	  ref <- bfcrpath(bfc, full.url)
+
+	  load(ref, envir = environment())
```
The way to do it is this:

```r
env <- new.env()
load(ref, envir=env)
ref.set <- get(dataset, envir=env)
```

LTLA commented on R/helper.R:
```diff
@@ -0,0 +1,43 @@
+#'       \item "immgen"
+#'       \item "mouse.rnaseq"
+#'     }
+#' @return SingleR reference dataset.
```
This is a bit too vague; I believe the output is a list with various fields of interest? Have a look at `trainSingleR`'s `@return` field for the level of detail required; think of what a skillful newcomer would need to interpret the output.

LTLA commented on R/helper.R:
```diff
@@ -0,0 +1,43 @@
+	  base.url <- "https://github.com/dviraran/SingleR/blob/master/data/"
+	  url.tail <- ".rda?raw=true"
+
+	  full.url <- paste0(base.url, dataset, url.tail)
```
This is fine, but you could also consider the more elegant:

```r
full.url <- sprintf("https://github.com/dviraran/SingleR/blob/master/data/%s.rda?raw=true", dataset)
```

LTLA commented:
Also, perhaps the name `getSinglerRef` is not ideal, for various reasons:

1.  It's already in the `SingleR` package, so there's no need to have `SingleR` in the name as well.
2.  `Singler` itself is misspelt to fit into the camelCase but this is a bit awkward.

Suggest maybe something like `getReferenceDatasets`.

> I'm willing to add a note about it to the vignette if you'd like.

Yes, the vignette has a stub usage section for the reference datasets. If there's one that works well with the pancreas datasets that I've currently got in there, that would be good to demonstrate.

j-andrews7 commented:
Latest commit should have addressed all of your comments, I think. My code in the vignette takes ~90 seconds, which I'm guessing pushes the vignette pretty close to the limit.

Let me know if you see any other issues.

LTLA commented:
Looks good. Would it help to subset the data in the vignette? Or is the bottleneck not related to the number of input cells?

j-andrews7 commented:
I didn't do any real benchmarking or debugging to test, but my guess is
that it's because the reference set is 700+ samples. And using the
pre-defined `de.genes` is a no-go, as deriving the common set likely
invalidates many of them. And I didn't super relish going through the lists
of lists to snag only those remaining.

If it's necessary, I'm sure there are ways to speed it up (including
subsetting the data - probably the simplest solution). There are ~1800 in
that set; subsetting to 500 would probably cut it down to a more manageable
level.

On Sun, Aug 4, 2019 at 12:24 AM Aaron Lun <notifications@github.com> wrote:

> Looks good. Would it help to subset the data in the vignette? Or is the
> bottleneck not related to the number of input cells?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/LTLA/SingleR/pull/5?email_source=notifications&email_token=ACOAQNDXE7IRXP3DUCJXKU3QCZRYBA5CNFSM4IJBD7ZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3P2W7Y#issuecomment-517974911>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOAQNCGXUMO3J3APEIRVOTQCZRYBANCNFSM4IJBD7ZA>
> .
>

LTLA commented:
Good enough, I'll sort it out at some point.

LTLA commented:
Oh, I should have told you to add yourself as a contributor. I can do that if that's fine with you.
